### DISCUSSION TITLE
Symfony 5.4 System Cache in Mult-Tenant Environment

### DISCUSSION THREAD

**Discussion by troy-rudolph** [symfony/symfony#52621] [Q&A]
Hello.  Symfony is relatively new to me.  We use Symfony 5.4 for our rather large RESTful API.   My question is about how to "manage" the Symfony Cache.

We have a very large multi-client (multi-tenant) application.  Like many apps, it is spread across 10 VMs (webheads) running Apache 2.4 with a load-balancer in front.  All ten webheads mount, via NFS, a common file system containing the files needed to execute and configure the app.  All the clients (1000+) run the same app and same version of the app.

We use the default adapter for the Symfony system cache.  Symfony is configured so that there is a Symfony system cache for each client (tenant).  We do not use any other Symfony caching features.

When we deploy a new release (package) of our application, the last step of the deployment process is to "rm -rf" the System symfony cache for each client.  As you might imagine, for 1000+ clients, this can take a long while (four hours).  We're looking to eliminate that deployment lag.

We did make an attempt at having all the clients share the Symfony system cache.  But in a busy app like ours, that led to a very troublesome cache stampede (dog pile) after clearing it.

1) Do we really need to clear the Symfony system cache after deploying a new version of the app? (I have done some local testing it seems to work fine without it.  However, it was limited testing by hand and I may missed something.)

2) If is necessary to clear the cache what is the most efficient way to do it?

3) Is it possible to have all 1000+ clients share the Symfony system cache?  If so, how do we handle the cache stampede?

4) What precisely does "cache warming" do?  I know it adds stuff to the cache.  From testing, however, it appears that most of it is adding an enormous number of folders.

5) Any and all suggestions are welcome.

**Comment by ghost** [comment]
`rm -rf` takes 4 hours 😮😮😮? What is an average size of the cache per client?

What do you mean by `Symfony Cache`? The HttpCache class?

1. The `cache` is a very general term. You can have different caches, different adapters.
Theoretically you don't have to clear the cache when deploying changes that don't affect the cache. However this would be very hard to keep track of every single change in the codebase and remember whether the cache has to be clearer or not. 
2. If your use filesystem cache mounted on NFS I would try to rotate the NFS every deploy. During deploy you mount the second one that is empty and then you can `rm -rf` the first one.  
3. It depends whether a cached content is client specific.

**Reply by troy-rudolph** [reply]
Our cache configuration is completely the default one.  So, as noted above, the default adapter (cache.adapter.system) is the one in use.  The build cache is also in its location.  So <path>/prod.cache contains the generated Symfony container files.  <path>/prod/cache/pools/system contains the Symfony system cache.

I admit that I am not sure what the class name of the adapter is.  It does not appear to be documented.  Also I admit again that this is all fairly new to me.  I did try to use the same terminology as the documentation however.

**Reply by ghost** [reply]
I think this cache can be shared because it's generated based on static PHP/yaml files. If it's generation depended on the client then it couldn't be shared.
But I still don't understand how this can take 4 hours to remove. Either you app is sooooo huge that it contains tons of files that are cached or you're not telling us the whole story.
